apps/mymedicare_cb/models.py

```diff
@@ -61,71 +61,92 @@ def get_and_update_user(
         else:
-            # Previously stored value was None/Null and mbi_hash != None, update just the mbi hash.
+            # Did the mbi change from previously stored None/Null value?
             if mbi_hash is not None:
-                err_msg = "UPDATE mbi_hash since previous value was NULL"
-                log_dict["status"] = "OK"
-                log_dict["mesg"] = err_msg
-                logger.info(log_dict)
-                user.crosswalk.user_mbi_hash = mbi_hash
+                mbi_updated = True
+                mbi_updated_from_null = True
+
+        # Update Crosswalk if there are any allowed changes or hash_type used for lookup changed.
+        if user.crosswalk.user_id_type != hash_lookup_type or hicn_updated or mbi_updated:
+            # Log crosswalk before state
+            log_dict.update({
+                "crosswalk_before": {
+                    "id": user.crosswalk.id,
+                    "user_hicn_hash": user.crosswalk.user_hicn_hash,
+                    "user_mbi_hash": user.crosswalk.user_mbi_hash,
+                    "fhir_id": user.crosswalk.fhir_id,
+                    "user_id_type": user.crosswalk.user_id_type,
+                },
+            })
+
+            with transaction.atomic():
+                # Archive to audit crosswalk changes
+                ArchivedCrosswalk.create(user.crosswalk)
+
+                # Update crosswalk per changes
                 user.crosswalk.user_id_type = hash_lookup_type
+                user.crosswalk.user_hicn_hash = hicn_hash
+                user.crosswalk.user_mbi_hash = mbi_hash
                 user.crosswalk.save()
 
-        # Update hash type used for lookup, if it has changed from last match.
-        if user.crosswalk.user_id_type != hash_lookup_type:
-            err_msg = (
-                "UPDATE user_id_type as it has changed from the previous lookup value"
-            )
-            log_dict["status"] = "OK"
-            log_dict["mesg"] = err_msg
-            logger.info(log_dict)
-            user.crosswalk.user_id_type = hash_lookup_type
-            user.crosswalk.save()
-
-        err_msg = "RETURN existing beneficiary record"
-        log_dict["status"] = "OK"
-        log_dict["mesg"] = err_msg
+        # Beneficiary has been successfully matched!
+        log_dict.update({
+            "status": "OK",
+            "user_id": user.id,
+            "user_username": user.username,
+            "hicn_updated": hicn_updated,
+            "mbi_updated": mbi_updated,
+            "mbi_updated_from_null": mbi_updated_from_null,
+            "mesg": "RETURN existing beneficiary record",
+            "crosswalk": {
+                "id": user.crosswalk.id,
+                "user_hicn_hash": user.crosswalk.user_hicn_hash,
+                "user_mbi_hash": user.crosswalk.user_mbi_hash,
+                "fhir_id": user.crosswalk.fhir_id,
+                "user_id_type": user.crosswalk.user_id_type,
+            },
```

Comment:
Moved the location of this logging code to show the final state of the `Crosswalk`. `crosswalk_before` was added to show the state before updates are made.